pkg/gatewayserver/io/ws/format.go

```diff
@@ -0,0 +1,48 @@
+// Copyright © 2020 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ws
+
+import (
+	"context"
+	"time"
+
+	"go.thethings.network/lorawan-stack/v3/pkg/gatewayserver/io"
+	"go.thethings.network/lorawan-stack/v3/pkg/gatewayserver/scheduling"
+	"go.thethings.network/lorawan-stack/v3/pkg/ttnpb"
+)
+
+// EndPoint contains information on the WebSocket endpoint.
+type EndPoint struct {
+	Scheme  string
+	Address string
+	Prefix  string
+}
+
+// Formatter abstracts messages to/from websocket based gateways.
+type Formatter interface {
+	// Connect creates necessary session state for a gateway.
+	Connect(ctx context.Context, uid string) error
+	// Disconnect removes the session state for a gateway.
+	Disconnect(ctx context.Context, uid string)
```

Comment:
If the formatter implementation is stateful, it's probably better to manage the state via construction (`New() (Formatter, error)`) and closure (`Close() error`).

Otherwise, it's unclear for the caller if `Connect()` can be used concurrently for the same instance of `Formatter` they have, or if there should be new formatters constructed.